File: bench/testdata/samples/A03/PSB-A03-CMD-001/go/vulnerable.go
// PSB-A03-CMD-001: Deployment script runner
// CWE: CWE-78
// Expected: BATOU-INJ-005, BATOU-INJ-006
package main

import (
	"encoding/json"
	"fmt"
	"net/http"
	"os/exec"
)

func deploy(w http.ResponseWriter, r *http.Request) {
	var req struct {
		Service     string `json:"service"`
		Environment string `json:"environment"`
	}
	json.NewDecoder(r.Body).Decode(&req)

	if req.Environment != "staging" && req.Environment != "production" {
		http.Error(w, `{"error":"invalid environment"}`, http.StatusBadRequest)
		return
	}

	cmd := fmt.Sprintf("/opt/deploy/run.sh %s %s", req.Service, req.Environment)
	out, err := exec.Command("sh", "-c", cmd).Output()
	if err != nil {
		http.Error(w, `{"error":"deployment failed"}`, http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]string{
		"status": "deployed",
		"output": string(out),
	})
}

File: bench/testdata/samples/A03/PSB-A03-CMD-001/javascript/vulnerable.js
// PSB-A03-CMD-001: Deployment script runner
// CWE: CWE-78
// Expected: GTSS-INJ-005, GTSS-INJ-006
const express = require("express");
const { exec } = require("child_process");
const app = express();
app.use(express.json());

app.post("/api/deploy", (req, res) => {
  const { service, environment } = req.body;

  if (!["staging", "production"].includes(environment)) {
    return res.status(400).json({ error: "invalid environment" });
  }

  const cmd = `/opt/deploy/run.sh ${service} ${environment}`;
  exec(cmd, (err, stdout, stderr) => {
    if (err) {
      return res.status(500).json({ error: "deployment failed", stderr });
    }
    res.json({ status: "deployed", output: stdout });
  });
});

module.exports = app;
